src/fe_utils/frontend_hook.c

```diff
@@ -0,0 +1,130 @@
+#include "fe_utils/frontend_hook.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <dlfcn.h>    // for dlopen()
+#include <sys/stat.h> // for stat()
+
+#include "c.h"
+#include "port.h"
+
+#define MAXPATHLEN 256 // in src/port/glob.c
+
+char pg_install_path[MAXPATHLEN] = "\0", pg_data_path[MAXPATHLEN] = "\0";
+
+static void *dlhandle[128] = {0};
+static int dlhandle_size = 0;
+
+// find which libraries need to load
+static void fronted_load_library(const char *full_path) {
+	struct stat stat_buf;
+	if (lstat(full_path, &stat_buf) != 0)
+	{
+		printf("lstat: %s\n", strerror(errno));
+		exit(-1);
+	}
+
+	void *h = dlopen(full_path, RTLD_NOW | RTLD_GLOBAL);
+	if (h == NULL)
+	{
+		char *error = dlerror();
+		printf("dlopen: %s\n", error);
+		exit(-1);
+	}
+
+	if (dlhandle_size >= 127)
+	{
+		printf("too many libraries to load");
+		exit(-1);
+	}
+
+	for (int i = 0; i < dlhandle_size; ++i)
+	{
+		if (dlhandle[i] == h)
+		{
+			return; // dll loaded, pass
+		}
+	}
+
+	dlhandle[dlhandle_size++] = h;
+
+	void (*f)(void) = dlsym(h, "_PG_init");
+	if (f == NULL)
+	{
+		printf("dlopen: can not find _PG_init() in %s\n", full_path);
+		exit(-1);
+	}
+
+	f();
+
+	return;
+}
+
+static void pg_find_env_path() {
+	char *install_path = NULL, *data_path = NULL;
+
+	if (install_path == NULL)
+	{
+		install_path = getenv("GPDB_DIR");
+	}
+	if (data_path == NULL)
+	{
+		data_path = getenv("MASTER_DATA_DIRECTORY");
```

Comment:
For frontend tools, some of them supports using the environment `PGDATA`, some of them supports using `pgdata`. It may cause users confusion when they have specified the `--pgdata` argument but the front-end hook requires them set up the `COORDINATOR_DATA_DIRECTORY`. What about initializing `data_path` at clients side after the client tool has parsed environment variables and its command line options?

Reply:
> What about initializing data_path at clients side after the client tool has parsed environment variables and its command line options?

https://github.com/Sasasu/gpdb/pull/10/files#diff-31ab3b8401a098ca86cb4df9ab209a8b77dad0ce445f1a035b642d92b0941147R530-R532

> I have strong feeling that TDE frontend needs PGDATA only if the utility needs it.

this is impossible, TDE need the file name, but PostgreSQL does not need like pg_checksum.

Reply:
We had an offline discussion that use a global var like `pg_xxx_data_dir` would be better than passing it by function args in this case. And client should set the var if it can, the client set value takes priority.